src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -115,30 +112,19 @@ public Task<string> ReadAsStringAsync()
                 // then check for a byte-order-mark (BOM) in the data to figure out the encoding.
                 if (encoding == null)
                 {
-                    byte[] preamble;
-                    foreach (Encoding testEncoding in s_encodingsWithBom)
-                    {
-                        preamble = testEncoding.GetPreamble();
-                        if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        {
-                            encoding = testEncoding;
-                            bomLength = preamble.Length;
-                            break;
-                        }
-                    }
+                    TryDetectEncoding(data, dataLength, ref encoding, ref bomLength);
                 }
 
-                // Use the default encoding if we couldn't detect one.
-                encoding = encoding ?? DefaultStringEncoding;
-
-                // BOM characters may be present even if a charset was specified.
-                if (bomLength == -1)
+                if (encoding == null)
                 {
-                    byte[] preamble = encoding.GetPreamble();
-                    if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        bomLength = preamble.Length;
-                    else
-                        bomLength = 0;
+                    // Use the default encoding if we couldn't detect one.
+                    encoding = DefaultStringEncoding;
+                    bomLength = 0;
+                }
+                else if (bomLength == -1)
```

Comment:
Or, move the `DefaultStringEncoding` fallback directly inside `TryDetectEncoding` and keep it void (in which case I'd probably rename it just `DetectEncoding`).

```c#
// If no content encoding is listed in the ContentType HTTP header, or no Content-Type header present,
// then check for a BOM in the data to figure out the encoding with fallback to DefaultStringEncoding.
if (encoding == null)
{
    DetectEncoding(data, dataLength, ref encoding, ref bomLength);
}
```